psc-publish/ErrorsWarnings.hs

```diff
@@ -247,32 +252,39 @@ displayOtherError e = case e of
     successivelyIndented
       [ "An IO exception occurred:", show exc ]
 
-renderWarnings :: [PackageWarning] -> Box
-renderWarnings =
-  collectWarnings
-    [ (getResolutionNotVersion, warnResolutionNotVersions)
-    , (getUndeclaredDependency, warnUndeclaredDependencies)
-    ]
-  where
-  collectWarnings patterns warns =
-    let boxes = mapMaybe (collectWarnings' warns) patterns
-        result = vcat
-                    [ para "Warnings:"
-                    , indented (vcat (intersperse spacer boxes))
-                    ]
-    in if null boxes then nullBox else result
+data CollectedWarnings = CollectedWarnings
```

Comment:
I wanted to collect all the warnings of each type up together, so that I can then create one `Box` for each type of warning, in order to make the messages easier to understand. This code makes sure that, for example, all of the `UndeclaredDependency` warnings are collected up and displayed together, giving a message like "the following are installed but not declared: [list]". It works similarly for the other two `PackageWarning` constructors.